### DISCUSSION TITLE
MDX blog just displays markdown content instead of rendering it while using `MDXProvider` from `@mdx-js/react` in Next JS

### DISCUSSION THREAD

**Discussion by deadcoder0904** [mdx-js/mdx#1276] [Q&A] [answered]
I am making a blog with [MDX](https://mdxjs.com) & [Next.js](https://nextjs.org) but I am unable to render Markdown content. The blog post just shows markdown content as `string`.

Here's my complete source code → https://github.com/deadcoder0904/blog-mdx-next/

I have the following folder structure:

```bash
.
|-- README.md
|-- components
|   `-- Image.js
|-- next.config.js
|-- package-lock.json
|-- package.json
|-- pages
|   |-- _app.js
|   |-- blog
|   |   `-- [slug].js
|   |-- dark.css
|   |-- index.js
|   `-- new.css
|-- posts
|   |-- blog
|   |   |-- hello-world
|   |   |   |-- Rustin_Cohle.jpg
|   |   |   `-- index.mdx
|   |   `-- shit-world
|   |       `-- index.mdx
|   `-- tutorials
|       `-- console-log-in-javascript
|           `-- index.mdx
|-- prettier.config.js
`-- utils
    `-- mdxUtils.js
```

I have all my content in `posts/` folder.

I have 2 folders in it: `blog/` & `tutorials/`

Each post is in their own folder inside of `blog/` or `tutorials/`. It looks like:

[![blog archives][1]][1]

I want to render my `posts/blog/hello-world/index.mdx` post at `blog/hello-world` location so I made `blog/[slug].js` file.

The contents of it are as follows:

# pages/blog/[slug].js

```js
import fs from 'fs'
import path from 'path'
import matter from 'gray-matter'
import { MDXProvider } from '@mdx-js/react'

import { BLOG_PATH, blogFilePaths } from '../../utils/mdxUtils'
import { Image } from '../../components/Image'

const MDXComponents = { Image }

const Blog = ({ source, frontMatter }) => {
	return (
		<div>
			<h1>{frontMatter.title}</h1>
			<MDXProvider components={MDXComponents}>{source}</MDXProvider>
		</div>
	)
}

export async function getStaticPaths() {
	const paths = blogFilePaths.map((path) => {
		const split = path.split('/')
		const slug = split[split.length - 2]
		return {
			params: {
				slug,
			},
		}
	})

	return {
		paths,
		fallback: false,
	}
}

export const getStaticProps = async ({ params }) => {
	const { slug } = params
	const blogFilePath = path.join(BLOG_PATH, `/blog/${slug}/index.mdx`)

	const source = fs.readFileSync(blogFilePath)
	const { content: mdx, data } = matter(source)

	if (!blogFilePath) {
		console.warn('No MDX file found for slug')
	}

	return {
		props: {
			source: mdx,
			frontMatter: data,
		},
	}
}

export default Blog
```

The important part is:

```js
<MDXProvider components={MDXComponents}>{source}</MDXProvider>
```

I thought this would render markdown content but it only displays markdown content as a `string`.

You can check the output at https://codesandbox.io/s/github/deadcoder0904/blog-mdx-next?file=/pages/blog/%5Bslug%5D.js by clicking any blog post.

It displays the following when I click on `Hello World` post:

[![hello world][2]][2]

How do I actually render the content?

I looked at other Github repos like [Tailwind CSS Blog](https://github.com/tailwindlabs/blog.tailwindcss.com) & it works [fine for them](https://blog.tailwindcss.com/) but I am not sure how it works sadly :(

I do know that I have to convert the `source` prop in `Blog` or `mdx` in `getStaticProps` but I [don't see](https://github.com/tailwindlabs/blog.tailwindcss.com/blob/master/src/components/Post.js#L88) Tailwind doing it either.

  [1]: https://i.stack.imgur.com/2X5bE.png
  [2]: https://i.stack.imgur.com/E2BJc.png

**Comment by deadcoder0904** [comment, marked as answer]
Made this work (technically used another solution), check branch `tailwind` :)

**Comment by mcmoodoo** [comment]
wow hah it's been almost 5 years an I am facing the exact same issue: nextjs, tailwind, next-mdx-remote...

I am surprised nobody even commented here in the last 4 years :)

**Reply by deadcoder0904** [reply]
well, now you did :)

**Reply by mcmoodoo** [reply]
Yes, so what's the current solution you're using? perhaps things have changed with mdx in nextjs...
